File: 0152-maximum-product-subarray/0152-maximum-product-subarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxProduct = function(nums) {
    let min = nums[0]
    let max = nums[0]
    let res = nums[0]
    for (let i=1; i < nums.length; i++) {
        const temp = max
        max = Math.max(nums[i], max * nums[i], min * nums[i])
        min = Math.min(nums[i], temp * nums[i], min * nums[i])
        res = Math.max(res, max)
    }
    return res
};

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        
        min_so_far = max_so_far = res = nums[0]
        
        for i in range(1, len(nums)):
            temp = max_so_far
            
            max_so_far = max(nums[i], max_so_far * nums[i], min_so_far * nums[i])
            min_so_far = min(nums[i], temp * nums[i], min_so_far * nums[i])
            res = max(res, max_so_far)
        
        return res